File: src/components/stepforms/schemas/evolution.ts
import { addNotInColumnNamesConstraint, StepFormType } from './utils';

const schema = {
  $schema: 'http://json-schema.org/draft-07/schema#',
  title: 'Evolution step',
  type: 'object',
  properties: {
    name: {
      type: 'string',
      consnt: 'evolution',
    },
    dateCol: {
      type: 'string',
      minLength: 1,
      title: 'Date column',
      description: 'Date column used to compute evolution',
      attrs: {
        placeholder: 'Select a column',
      },
    },
    valueCol: {
      type: 'string',
      minLength: 1,
      title: 'Value column',
      description: 'The value column',
      attrs: {
        placeholder: 'Select a column',
      },
    },
    evolutionType: {
      type: 'string',
      title: 'Compute evolution versus',
      enum: ['vsLastYear', 'vsLastMonth', 'vsLastWeek', 'vsLastDay'],
    },
    evolutionFormat: {
      type: 'string',
      title: 'Compute evolution in',
      enum: ['abs', 'pct'],
    },
    indexColumns: {
      type: 'array',
      minItems: 0,
      items: {
        type: 'string',
        minLength: 1,
      },
      title: 'indexColumns',
      description: 'Index columns',
      attrs: {
        placeholder: 'Select columns',
      },
    },
    newColumn: {
      type: 'string',
      minLength: 1,
      title: 'New evolution column',
      description: 'New column to be created',
    },
  },
  required: ['name', 'dateCol', 'valueCol', 'evolutionType', 'evolutionFormat', 'indexColumns'],
  additionalProperties: false,
};

export default function buildSchema(form: StepFormType) {
  return addNotInColumnNamesConstraint(schema, 'newColumn', form.columnNames);
}

File: server/src/weaverbird/pipeline/steps/evolution.py
from typing import Literal

from pydantic import Field

from weaverbird.pipeline.steps.utils.base import BaseStep
from weaverbird.pipeline.steps.utils.render_variables import StepWithVariablesMixin
from weaverbird.pipeline.types import TemplatedVariable

EVOLUTION_TYPE = Literal["vsLastYear", "vsLastMonth", "vsLastWeek", "vsLastDay"]
EVOLUTION_FORMAT = Literal["abs", "pct"]


class EvolutionStep(BaseStep):
    name: Literal["evolution"] = "evolution"
    date_col: str
    value_col: str
    evolution_type: EVOLUTION_TYPE
    evolution_format: EVOLUTION_FORMAT
    index_columns: list[str] = Field(default_factory=list)
    new_column: str | None = None


class EvolutionStepWithVariable(EvolutionStep, StepWithVariablesMixin):
    index_columns: TemplatedVariable | list[TemplatedVariable] = Field(default_factory=list)
